src/messageschannel.cpp

```diff
@@ -90,14 +92,64 @@ MatrixMessagesChannel::MatrixMessagesChannel(MatrixConnection *connection, Tp::B
     }
 }
 
+void MatrixMessagesChannel::onPendingEventChanged(int pendingEventIndex)
+{
+    // Delivery Report message
+    // https://telepathy.freedesktop.org/spec/Channel_Interface_Messages.html#Enum:Delivery_Status
+    // https://matrix.org/docs/spec/client_server/r0.4.0.html#put-matrix-client-r0-rooms-roomid-send-eventtype-txnid
+
+    const QMatrixClient::PendingEventItem &pendingEvent = m_room->pendingEvents().at(pendingEventIndex);
+    Tp::DeliveryStatus tpDeliveryStatus;
+    switch (pendingEvent.deliveryStatus()) {
+    case QMatrixClient::EventStatus::ReachedServer:
+        tpDeliveryStatus = Tp::DeliveryStatusAccepted;
+        break;
+    case QMatrixClient::EventStatus::SendingFailed:
+        tpDeliveryStatus = Tp::DeliveryStatusTemporarilyFailed;
+        break;
+    default:
+        tpDeliveryStatus = Tp::DeliveryStatusUnknown;
+        break;
+    }
+
+    Tp::MessagePartList partList;
+
+    Tp::MessagePart header;
+    header[QStringLiteral("message-sender")]    = QDBusVariant(m_targetHandle);
+    header[QStringLiteral("message-sender-id")] = QDBusVariant(m_targetId);
+    header[QStringLiteral("message-type")]      = QDBusVariant(Tp::ChannelTextMessageTypeDeliveryReport);
+    header[QStringLiteral("delivery-status")]   = QDBusVariant(tpDeliveryStatus);
+    header[QStringLiteral("delivery-token")]    = QDBusVariant(pendingEvent.event()->transactionId());
+    partList << header;
+
+    addReceivedMessage(partList);
+}
+
 MatrixMessagesChannelPtr MatrixMessagesChannel::create(MatrixConnection *connection, Tp::BaseChannel *baseChannel)
 {
     return MatrixMessagesChannelPtr(new MatrixMessagesChannel(connection, baseChannel));
 }
 
 QString MatrixMessagesChannel::sendMessageCallback(const Tp::MessagePartList &messageParts, uint flags, Tp::DBusError *error)
 {
-    return QString();
+    if (!m_room) {
```

Comment:
This is not possible (once the staged commits will meet the master branch). Now we always have valid m_room.